### PR TITLE
batcheval: add ComputeStatsDiff arg to addsstable

### PR DESCRIPTION
**Pull request by msbutler**
ComputeStatsDiff causes the server to compute the effect this sst will have on
the range's mvcc stats, even in the presence of overlapping keys. This flag
cannot be passed with the MVCCStats, DisallowShadowingBelow, or
DisallowShadowing fields.

This flag assumes that any key in the sst that is shadowed by a key in the
engine is also a duplicate. As an example, accurate stats will be computed
here:

 - sst: a@3,a@2 and eng: a@2,a@1
 but not for, as a@1 is not a duplicate:
 - sst: a@1 and eng: a@2

Informs https://github.com/cockroachdb/cockroach/issues/145548

Release note: none

### CODE REVIEW

**Comment by msbutler** [comment]
unrelated test flake

File: pkg/kv/kvserver/batcheval/cmd_add_sstable.go [hunk 1]
```diff
@@ -250,112 +260,65 @@ func EvalAddSSTable(
...
+	}
+	if args.ComputeStatsDiff {
+		if checkConflicts {
+			return result.Result{}, errors.New(
+				"AddSSTableRequest.ComputeStatsDiff cannot be used with DisallowConflicts or DisallowShadowingBelow")
+		}
+		if args.MVCCStats != nil {
+			return result.Result{}, errors.New(
+				"AddSSTableRequest.ComputeStatsDiff cannot be used with precomputed MVCCStats")
 		}
```
Comment (stevendanna) [inline pkg/kv/kvserver/batcheval/cmd_add_sstable.go]: I've started implementing a `Validate` method on request objects (see `pkg/kv/kvpb/api.go`) for these kind of checks that only depend on data in the request or the header.  Mind doing the same for this request and then moving this check up before we've potentially done some work.

File: pkg/kv/kvpb/api.proto [hunk 2]
```diff
@@ -2112,6 +2112,19 @@ message AddSSTableRequest {
   //
   // TODO(dt,msbutler,bilal): This is unsupported.
   util.hlc.Timestamp ignore_keys_above_timestamp = 12 [(gogoproto.nullable) = false];
+  
+  // ComputeStatsDiff causes the server to compute the effect this
+  // sst will have on the range's mvcc stats, even in the presence of
+  // overlapping keys. This flag cannot be passed with the MVCCStats,
+  // DisallowShadowingBelow, or DisallowShadowing fields.
```
Comment (stevendanna) [inline pkg/kv/kvpb/api.proto]:
```suggestion
  // ComputeStatsDiff causes the server to compute the effect this
  // SST will have on the range's MVCC stats, even in the presence of
  // overlapping keys. This flag cannot be passed with the MVCCStats,
  // DisallowShadowingBelow, or DisallowShadowing fields. 
```

File: pkg/kv/kvpb/api.proto [hunk 3]
```diff
@@ -2112,6 +2112,19 @@ message AddSSTableRequest {
...
+  // sst will have on the range's mvcc stats, even in the presence of
+  // overlapping keys. This flag cannot be passed with the MVCCStats,
+  // DisallowShadowingBelow, or DisallowShadowing fields. 
+  //
+  // This flag assumes that any key in the sst that is shadowed by a key in the
+  // engine is also a duplicate. As an example, accurate stats will be computed
+  // here: 
+  // - sst: a3,a2 and eng: a2,a1
+  // but not for, as a1 is not a duplicate: 
+  // - sst: a1 and eng: a2
```
Comment (stevendanna) [inline pkg/kv/kvpb/api.proto]: Can we say what happens in this case? Is an error returned? Do we silently create inaccurate stats?

In other places we sometimes use `a@1` to indicate "Key A with Timestamp 1".

Aside: For PCR, I think the only way this would happen is if GC was allowed on the range we are ingesting into. Since we have a protected timestamp on the whole tenant that only advances with the resolved timestamp, I _think_ we should be pretty safe from this edge case.

What we could consider is some test only code (perhaps in or around `EarliestActiveTimestamp`) that enforces that any AddSSTable without `SSTTimestampToRequestTimestamp` set can't contain a key with a timestamp below the gc threshold of the range.   Not for this PR, but if we are particularly worried here it might provide us a way to discover cases we haven't thought of.

File: pkg/kv/kvserver/batcheval/cmd_add_sstable.go [hunk 4]
```diff
@@ -250,112 +260,65 @@ func EvalAddSSTable(
...
+		statsDiff, err := computeSSTStatsDiffWrapper(ctx, sst, readWriter, h.Timestamp.WallTime, start, end)
+		if err != nil {
+			return result.Result{}, errors.Wrap(err, "computing SST stats diff")
+		}
+		stats.Add(statsDiff)
+	} else if args.MVCCStats != nil {
+		stats.Add(*args.MVCCStats)
 	} else {
-		stats.ContainsEstimates++
+		sstStats, err := computeSSTStats(ctx, sst, h.Timestamp.WallTime)
```
Comment (stevendanna) [inline pkg/kv/kvserver/batcheval/cmd_add_sstable.go:310]: I found my first read through this code a little confusing. I think it _might_ be helped by naming the option `ComputeAccurateStats` and then renaming this function `computeSSTStatsEstimate` or something along those lines.

File: pkg/kv/kvserver/batcheval/cmd_add_sstable.go [hunk 5]
```diff
@@ -250,112 +260,65 @@ func EvalAddSSTable(
...
-		if statsDelta.ContainsEstimates == 0 {
-			stats.ContainsEstimates = 0
+		stats.Add(checkConflictsStatsDelta)
+	}
+	if !(checkConflicts || args.ComputeStatsDiff) {
+		// If CheckSSTConflicts or ComputeStatsDiff are not used to compute stats,
+		// then the stats do not account for overlapping keys in the engine, so we
+		// have to assume they are estimates.
+		stats.ContainsEstimates++
+	}
```
Comment (stevendanna) [inline pkg/kv/kvserver/batcheval/cmd_add_sstable.go]: I think it might be clearer to duplicate a little code and directly do this `ContainsEstimates++ on the branches below that need it. But I don't feel strongly on this one.

File: pkg/kv/kvserver/batcheval/cmd_add_sstable.go [hunk 6]
```diff
@@ -505,6 +468,89 @@ func EvalAddSSTable(
 	}, nil
 }
 
+func computeSSTStatsDiffWrapper(
```
Comment (stevendanna) [inline pkg/kv/kvserver/batcheval/cmd_add_sstable.go]: Maybe `computeAccurateSSTStatsWithFallback`

File: pkg/kv/kvserver/batcheval/cmd_add_sstable.go [hunk 7]
```diff
@@ -505,6 +468,89 @@ func EvalAddSSTable(
...
+		if err != nil {
+			return enginepb.MVCCStats{}, errors.Wrap(err, "computing SSTable MVCC stats")
+		}
+		return sstStats, nil
+	}
+	return enginepb.MVCCStats{}, nil
+}
+
+// checkSSTSpanBounds verifies that the keys in the sstable are within the
+// span specified by the request header.
```
Comment (stevendanna) [inline pkg/kv/kvserver/batcheval/cmd_add_sstable.go]:
```suggestion
// checkSSTSpanBounds verifies that the keys in the sstable are within the
// span specified by [start, end).
```

File: pkg/kv/kvserver/batcheval/cmd_add_sstable.go [hunk 8]
```diff
@@ -194,10 +198,16 @@ func EvalAddSSTable(
 		}
 	}
 
-	var statsDelta enginepb.MVCCStats
 	maxLockConflicts := storage.MaxConflictsPerLockConflictError.Get(&cArgs.EvalCtx.ClusterSettings().SV)
 	targetLockConflictBytes := storage.TargetBytesPerLockConflictError.Get(&cArgs.EvalCtx.ClusterSettings().SV)
 	checkConflicts := args.DisallowConflicts || !args.DisallowShadowingBelow.IsEmpty()
+
+	// checkConflictsStatsDelta is a delta between the SST-only statistics and
+	// their effect when applied. In other words:
```
Comment (stevendanna) [inline pkg/kv/kvserver/batcheval/cmd_add_sstable.go:211]: Perhaps in another commit, I think this flow could be a bit clearer. One thing that obfuscates it currently is that the first "Add" of stats you see when reading through the code after this comment is:

```
checkConflictsStatsDelta.Add(sstReqStatsDelta)
```

at line 244. But that is the diff caused by the rewrite, which is a little confusing to me.

**Review by stevendanna — COMMENTED** [review]
Flushed my first pass. Overall, this seems like the right direction, I mostly have nitpics that may just be the result of me not being that familiar with the stats stuff.

File: pkg/kv/kvserver/batcheval/cmd_add_sstable.go (hunk 1 above)
Comment (msbutler) [reply pkg/kv/kvserver/batcheval/cmd_add_sstable.go]: Done.

File: pkg/kv/kvserver/batcheval/cmd_add_sstable.go (hunk 5 above)
Comment (msbutler) [reply pkg/kv/kvserver/batcheval/cmd_add_sstable.go]: Done.

File: pkg/kv/kvserver/batcheval/cmd_add_sstable.go (hunk 6 above)
Comment (msbutler) [reply pkg/kv/kvserver/batcheval/cmd_add_sstable.go]: hrm, I'm hesitant to rename this to include `accurate` since some call paths with `checkSSTConflicts` also compute accurate stats. After this merges, jeff and I plan audit all clients to see if we can remove the MVCC field from addsstable request. If we can remove that field, i think this control flow will simplify signficantly.

File: pkg/kv/kvserver/batcheval/cmd_add_sstable.go (hunk 4 above)
Comment (msbutler) [reply pkg/kv/kvserver/batcheval/cmd_add_sstable.go:310]: see comment below.

File: pkg/kv/kvserver/batcheval/cmd_add_sstable.go (hunk 6 above)
Comment (msbutler) [reply pkg/kv/kvserver/batcheval/cmd_add_sstable.go]: renamed to `computeSSTStatsDiffWithFallBack`

File: pkg/kv/kvserver/batcheval/cmd_add_sstable.go (hunk 7 above)
Comment (msbutler) [reply pkg/kv/kvserver/batcheval/cmd_add_sstable.go]: Done.

File: pkg/kv/kvserver/batcheval/cmd_add_sstable.go (hunk 8 above)
Comment (msbutler) [reply pkg/kv/kvserver/batcheval/cmd_add_sstable.go:211]: Yeah, once this merges, i'd like to clean up this logic.

File: pkg/kv/kvpb/api.proto (hunk 3 above)
Comment (msbutler) [reply pkg/kv/kvpb/api.proto]: > Can we say what happens in this case?

added a few more words.

> What we could consider is some test only code (perhaps in or around EarliestActiveTimestamp) that enforces that any AddSSTable without SSTTimestampToRequestTimestamp set can't contain a key with a timestamp below the gc threshold of the range. Not for this PR, but if we are particularly worried here it might provide us a way to discover cases we haven't thought of.

This is a nice idea! I'll add this to my todo list.

**Review by sumeerbhola — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147919)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson, @msbutler, and @stevendanna)

___
*[`pkg/kv/kvpb/api.proto` line 2123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147919#-OSz3EFX9FbLXW1Lua2U:-OSz3EFX9FbLXW1Lua2V:bjhowim) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4de1d12d8e08423b352e95e837eb297d5d1b664/pkg/kv/kvpb/api.proto#L2123)):*
> ```proto
>   // This flag assumes that any key in the sst that is shadowed by a key in the
>   // engine is also a duplicate. As an example, accurate stats will be computed
>   // here: 
> ```

(drive by)
If one of these exceptional conditions happens, and because we have disallowed the `MVCCStats` field from being populated, will we have no change to stats? Would it be better for the client to also populate the `MVCCStats` field, which would be treated as imperfect stats that will overestimate. Underestimating stats until stats recomputation happens seems risky, since it could result in a huge range.

**Review by msbutler — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147919)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson, @stevendanna, and @sumeerbhola)

___
*[`pkg/kv/kvpb/api.proto` line 2123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147919#-OSz3EFX9FbLXW1Lua2U:-OSz7pJ79LU1A0kKG7Uj:btxbd0d) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4de1d12d8e08423b352e95e837eb297d5d1b664/pkg/kv/kvpb/api.proto#L2123)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

(drive by)
If one of these exceptional conditions happens, and because we have disallowed the `MVCCStats` field from being populated, will we have no change to stats? Would it be better for the client to also populate the `MVCCStats` field, which would be treated as imperfect stats that will overestimate. Underestimating stats until stats recomputation happens seems risky, since it could result in a huge range.
</blockquote></details>

I'd like to leave the logic as is. Jeff and I would like to remove the MVCCStats field all together and conduct all stats computation in kvserver. 

In a follow PR, i'd like to detect this corner case and increment ContainsEstimates, as I describe here.  https://github.com/msbutler/cockroach/blob/butler-mvcc-ingest-stats/pkg/storage/sst.go#L225

**Review by msbutler — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147919)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson, @stevendanna, and @sumeerbhola)

___
*[`pkg/kv/kvpb/api.proto` line 2123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147919#-OSz3EFX9FbLXW1Lua2U:-OSz8nW0-gPwI0cxjRLS:b3024hb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4de1d12d8e08423b352e95e837eb297d5d1b664/pkg/kv/kvpb/api.proto#L2123)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

I'd like to leave the logic as is. Jeff and I would like to remove the MVCCStats field all together and conduct all stats computation in kvserver. 

In a follow PR, i'd like to detect this corner case and increment ContainsEstimates, as I describe here.  https://github.com/msbutler/cockroach/blob/butler-mvcc-ingest-stats/pkg/storage/sst.go#L225
</blockquote></details>

leaving this corner case around for now feels low risk as PCR will be the only user of this flag.

File: pkg/kv/kvserver/batcheval/cmd_add_sstable.go (hunk 6 above)
Comment (jeffswenson) [reply pkg/kv/kvserver/batcheval/cmd_add_sstable.go]: We won't be able to remove the field from the client until there we delete the current version from the supported upgrade chain (since the client may be sending requests to a server that does not support this option).

**Review by sumeerbhola — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147919)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson, @msbutler, and @stevendanna)

___
*[`pkg/kv/kvpb/api.proto` line 2123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147919#-OSz3EFX9FbLXW1Lua2U:-OSzH4KsAGBRo4PpN93D:bkv000l) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4de1d12d8e08423b352e95e837eb297d5d1b664/pkg/kv/kvpb/api.proto#L2123)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

leaving this corner case around for now feels low risk as PCR will be the only user of this flag.
</blockquote></details>

@stevendanna as the KV person on this review should give the authoritative opinion. The risks I see are:
- We wanted to do MVCCStats computation on the client when possible since that reduces time on the server, while holding latches (it also alleviated server hotspots when clients are fanning in). Given the number of options on AddSSTableRequest, it is hard for me to reconstruct what combinations are used when.
- Even if PCR is the only user of this flag, risking that a range will blow up to be huge seems inadvisable based on the observation that everything bad that is permissible can happen at scale.

File: pkg/kv/kvpb/api.proto [hunk 9]
```diff
@@ -2112,6 +2112,26 @@ message AddSSTableRequest {
...
+  //
+  // This flag assumes that any key in the sst that is shadowed by a key in the
+  // engine is also a duplicate. As an example, accurate stats will be computed
+  // here: 
+  // - sst: a@3,a@2 and eng: a@2,a@1
+  //
+  // but not for, as a@1 is not a duplicate: 
+  // - sst: a@1 and eng: a@2
+  //
+  // In the ladder case, we silently create inaccurate stats currently. A TODO
```
Comment (stevendanna) [inline pkg/kv/kvpb/api.proto:2129]:
```suggestion
  // In the latter case, we silently create inaccurate stats currently. A TODO
```

File: pkg/kv/kvpb/api.go [hunk 10]
```diff
@@ -2563,3 +2563,17 @@ func validateExclusionTimestampForBatch(ts hlc.Timestamp, h Header) error {
 	}
 	return nil
 }
+
+func (r *AddSSTableRequest) Validate(bh Header) error {
+	if r.ComputeStatsDiff {
+		if r.DisallowConflicts || r.DisallowShadowingBelow.IsSet() {
+			return errors.New(
```
Comment (stevendanna) [inline pkg/kv/kvpb/api.go:2570]: Perhaps `AssertionFailedf`

**Review by msbutler — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147919)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson and @stevendanna)

___
*[`pkg/kv/kvpb/api.proto` line 2123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147919#-OSz3EFX9FbLXW1Lua2U:-OT-6qs699LcJ_w8-29T:b-1ri0x8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4de1d12d8e08423b352e95e837eb297d5d1b664/pkg/kv/kvpb/api.proto#L2123)):*

> Even if PCR is the only user of this flag, risking that a range will blow up to be huge seems inadvisable
I've realized the PCR destination side GC race story we've been thinking about is largely academic: PCR holds a protected timestamp that lags 4 hours behind PCRs replicated time. Further, PCR ingests data with timestamps at or above the replicated time-- it will never ingest data with a timestamp over 4 hours behind its replicated time. So, the PTS will prevent GC over the time range PCR will be ingesting data for. 

Furthermore, the current PCR destination side mvcc stats situation is totally broken, as described in the parent issue.  So even if we're missing some corner case scenario, i think this patch leads to much better stats than without it.

**Review by msbutler — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147919)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson and @stevendanna)

___
*[`pkg/kv/kvpb/api.proto` line 2123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147919#-OSz3EFX9FbLXW1Lua2U:-OT-AjcpEHxMMVk683jo:b-xltn5w) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4de1d12d8e08423b352e95e837eb297d5d1b664/pkg/kv/kvpb/api.proto#L2123)):*
I had a reviewable formatting snafoo. here's  my response to sumeer, with better formatting.

>Even if PCR is the only user of this flag, risking that a range will blow up to be huge seems inadvisable


I've realized the PCR destination side GC race story we've been thinking about is largely academic: PCR holds a protected timestamp that lags 4 hours behind PCRs replicated time. Further, PCR ingests data with timestamps at or above the replicated time-- it will never ingest data with a timestamp over 4 hours behind its replicated time. So, the PTS will prevent GC over the time range PCR will be ingesting data for.

Furthermore, the current PCR destination side mvcc stats situation is totally broken, as described in the parent issue. So even if we're missing some corner case scenario, i think this patch leads to much better stats than without it.

**Comment by stevendanna** [comment]
> We wanted to do MVCCStats computation on the client when possible since that reduces time on the server, while holding latches (it also alleviated server hotspots when clients are fanning in). Given the number of options on AddSSTableRequest, it is hard for me to reconstruct what combinations are used when.

Perhaps I should be more worried about the latching, but since AddSSTableRequests are only really issued for keyspace related to offline tables or tenants, my main concern is just about the overall cost of this.

Now that we have this function, have we done any spot checks on what the performance looks like when this is wired up end-to-end?

To give some context to my concern here, the last time I thought about this was in the context of the change that led tot his comment: https://github.com/cockroachdb/cockroach/blob/d70ca10ca2d53f38a894d1c3c3f5697d76d024e4/pkg/backup/restore_data_processor.go#L502 

We saw that checking for conflicts dominated wall clock time of restore with even just a small amount of overlap.  But, a lot has changed since then, perhaps this is less costly now.

(As an aside, I'm still curious about whether PCR has properties that we could exploit in conjunction with a time-bound iterator to make this cheaper somehow, but that is for a different PR for sure.)

> Furthermore, the current PCR destination side mvcc stats situation is totally broken, as described in the parent issue. So even if we're missing some corner case scenario, i think this patch leads to much better stats than without it.

If I understand Sumeer correctly, we also want to think about the _direction_ of the error.  While the caller-generated statistics are inaccurate, they overestimate the range size (leading to over splitting).  The server-generated statistics, in these edge cases, under estimate the range size, which could lead to under splitting. The impact of a very large range in the system could be worse than the impact of over splitting.

**Review by stevendanna — APPROVED** [review]

File: pkg/kv/kvpb/api.proto (hunk 9 above)
Comment (dt) [reply pkg/kv/kvpb/api.proto:2129]: I'll get this in a follow-up.

File: pkg/kv/kvpb/api.go (hunk 10 above)
Comment (dt) [reply pkg/kv/kvpb/api.go:2570]: I'll get this in a follow-up.

**Comment by dt** [comment]
> The impact of a very large range in the system could be worse than the impact of over splitting.

Much worse. I think we'll need to figure this out before this can be default on, even just in PCR.

Per discussion with DR and with Steven elsewhere, we want to go ahead and merge this as-is now before we cut 25.3, and add metamorphic coverage of it, while we continue to test it out/refine it, rather than wait and then need to consider a backport later.

bors r=steven
